### DISCUSSION TITLE
how to get package information from python?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [pkgcore/pkgcore#439] [Q&A]
I am writing [a language server](https://github.com/termux/termux-language-server) for gentoo ebuild and hope to complete package name. For example, user input app-editors/n will complete app-editors/neovim and app-editors/nano . Is there a python API like get_ebuild_information('app-editors/n') and return a list like `[{"name": "app-editors/neovim", "version": "0.10.0", "description": "XXX"}, ...]`?

Not only pkgcore, portage or eix or gentoolkit or portageutils is OK if they provide enough API.

TIA!

https://forums.gentoo.org/viewtopic-p-8847461.html

**Comment by powerman** [comment]
I'm pretty sure there is a python API for this - because Gentoo portage itself is implemented in a python.
But I'm not a python developer, so to me there are other ways to get this info:
1. This command outputs list of all directories with Gentoo packages: `portageq get_repo_path / $(portageq get_repos /)`
2. Next look for list of subdirectories `app-editors/*/` inside each of directory from step 1.

**Comment by ferringb** [comment]
Pkgcore has a full API for package interaction.  For a starter, `pydoc pkgcore.ebuild.ebuild.base`; that shows the various metadata attributes exposed on an ebuild (whether source or binpkg).

For querying packages- pkgcore is built around the notion of 'restrictions' which represent search clauses (it's the same thing used for representing dependencies internally).  Start with `pkgcore.util` , specifically `pkgcore.util.parserestrict`.  The functionally that supports things like `pquery */diffball` or `pquery dev-util/*` is via that function.

The thing you need to be aware of for both portage and pkgcore is they internally cache their view of the ebuild tree.  IE, if you're mutating it under it's feet, you *do* have to force a reload (cache wipe) to see the changes.

Either way, getting a load config and the 'domain' (which representing a specific root + repo configuration):
```python
from pkgcore.config import load_config
domain = load_config()['livefs domain']
```
domain.repos gets you all repos that are sources of installation (ebuilds, binpkg, and various meta shit) w/ filtering applied- ACCEPT_LICENSE, ACCEPT_KEYWORDS, all of that.

For your needs, you probably want to start with `domain.all_repos_raw`.  That gets you something derived from `pkgcore.repository.prototype.tree`.  The part you care about for a 'tree' (a repo like object) is `.itermatch()`- feed it a restriction (search objects) and it'll return the packages that match.

**Reply by Freed-Wu** [reply]
Can you provide an example to match `app-editors/n` and return the information of `app-editors/neovim`, `app-editors/nano`?

TIA!

**Reply by ferringb** [reply]
```
>>> from pkgcore.config import load_config
>>> dom = load_config().get_default('domain')
>>> from pkgcore.util.parserestrict import parse_match
>>> dom.all_repos.match(parse_match("app-editors/nano"))
[<PackageWrapper pkg=<<class 'pkgcore.ebuild.ebuild_src.package'> cpv='app-editors/nano-8.0' @0x7f4012174590> wrapped='use' @0x7f40121e9950>, <PackageWrapper pkg=<<class 'pkgcore.ebuild.ebuild_src.package'> cpv='app-editors/nano-8.1' @0x7f4012174ad0> wrapped='use' @0x7f40121eaad0>, <PackageWrapper pkg=<<class 'pkgcore.ebuild.ebuild_src.package'> cpv='app-editors/nano-8.2' @0x7f4012174c20> wrapped='use' @0x7f40121eadd0>, <pkgcore.vdb.ondisk._WrappedInstalledPkg object at 0x7f40121cc750>]
```

you probably want to poke through what domain objects expose; `all_repos` exposes configuration wrapped repositories, collapsed into one queriable repo/tree.  The 'configuration wrapped' means USE configuration is injected in, as is license and keyword masking- if you want to see the *raw* repo, `all_repo_raw` .  For example:

```
>>> [x for x in dir(dom) if 'repo' in x and not x.startswith("_")]
['add_repo', 'all_binary_repos', 'all_binary_repos_raw', 'all_binary_repos_unfiltered', 'all_ebuild_repos', 'all_ebuild_repos_raw', 'all_ebuild_repos_unfiltered', 'all_installed_repos', 'all_installed_repos_raw', 'all_repos', 'all_repos_raw', 'all_source_repos', 'all_source_repos_raw', 'all_unfiltered_repos', 'binary_repos', 'binary_repos_raw', 'binary_repos_unfiltered', 'ebuild_repos', 'ebuild_repos_raw', 'ebuild_repos_unfiltered', 'filter_repo', 'find_repo', 'installed_repos', 'installed_repos_raw', 'repo_configs', 'repos', 'repos_raw', 'source_repos', 'source_repos_raw', 'unfiltered_repos']
```
Note that `parse_match` is *not* an EAPI atom parser- it's what is used for `pquery <some query>`.  IE, 'dev-util/*' works for that, as does stuff like '*/diffball' or 'diffball' (which it converts into '*/diffball').  If you want strict atom parsing, use this:
```
>>> from pkgcore.ebuild.atom import atom
>>> dom.all_repos.match(atom("app-editors/vim"))
[<PackageWrapper pkg=<<class 'pkgcore.ebuild.ebuild_src.package'> cpv='app-editors/vim-9.0.2167' @0x7f4012175010> wrapped='use' @0x7f4011f45dd0>, <PackageWrapper pkg=<<class 'pkgcore.ebuild.ebuild_src.package'> cpv='app-editors/vim-9.1.0366' @0x7f4012175160> wrapped='use' @0x7f4011f470d0>]
```

**Comment by Gao123861** [comment]
Gentoo's portage package manager provides a Python API to query package information. You can use the portage module's db interface to retrieve package information. Here is a simple example code that demonstrates how to use the portage module to retrieve package information for a specified category:
 
import portage
def get_ebuild_information(category, partial_name):
    db = portage.db[portage.root]["vartree"].dbapi
    matching_packages = db.match(f"{category}/{partial_name}*")
    results = []
    for cpv in matching_packages:
        pkg = db[cpv]
        results.append({
            "name": pkg.cpv,
            "version": pkg.version,
            "description": pkg.metadata["DESCRIPTION"]
        })
    return results

category = "app-editors"
partial_name = "n"
package_info = get_ebuild_information(category, partial_name)
print(package_info)
 This code first imports the portage module, and then defines a function called get_ebuild_information, which accepts a category and a component name as parameters and returns a list of package information that matches. The function uses the portage.db interface to query matching packages and extracts the name, version, and description information for each package.

**Reply by Freed-Wu** [reply]
It doesn't support glob:

> portage.exception.InvalidAtom: app-editors/n*

And db cannot be indexed.

```
In [22]: db["app-editors/neovim-0.10.2"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'vardbapi' object is not subscriptable
```

**Reply by ferringb** [reply]
> It doesn't support glob:
> 
> > portage.exception.InvalidAtom: app-editors/n*
> 
> And db cannot be indexed.
> 
> ```
> In [22]: db["app-editors/neovim-0.10.2"]
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: 'vardbapi' object is not subscriptable
> ```
That's a discussion for portage's upstream, but you're probably looking for `portage.dep.Atom(..., allow_wildcards=True)`.  `allow_repo=True` is probably of interest- if I'm reading this correctly, that'll flip on the '::<repo>' extended restriction.
